photon-ml/src/main/scala/com/linkedin/photon/ml/data/AvroDataReader.scala

```diff
@@ -0,0 +1,375 @@
+ *  but in the case of Photon, we still use partition counts in a few places to tune the performance of algorithms. The
+ *  DataFrame runtime has no knowledge of this and badly underestimates the number of partitions needed, which causes a
+ *  huge repartition/shuffle step during reading. Here we pass the partition count explicitly to the "read" function to
+ *  get around this problem. Once we revisit Photon's sensitivity to choice of partition count, we can remove this
+ *  parameter.
+ *
+ * @param sc the Spark context
+ * @param defaultFeatureColumn the default column to use for features
+ */
+class AvroDataReader(
+    sc: SparkContext,
+    defaultFeatureColumn: String = "features")
+  extends DataReader(defaultFeatureColumn) {
+
+  import AvroDataReader._
+
+  /**
+   * Internal sql context
+   */
+  private val sqlContext = new SQLContext(sc)
+
+  /**
+   * Reads the avro file at the given path into a DataFrame, generating a default index map for feature names. Merges
+   * source columns into combined feature vectors as specified by the featureColumnMap argument. Often features are
+   * joined from different sources, and it can be more scalable to combine them into problem-specific feature vectors
+   * that can be independently distributed.
+   *
+   * @param paths the paths to the avro files or folders
+   * @param featureColumnMap a map that specifies how the feature columns should be merged. The keys specify the name
+   *   of the merged destination column, and the values are sets of source columns to merge, e.g.:
+   *
+   *     Map("userFeatures" -> Set("profileFeatures", "titleFeatures"))
+   *
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @param numPartitions the minimum number of partitions. Spark is generally moving away from manually specifying
+   *   partition counts like this, in favor of inferring it. However, Photon currently still exposes partition counts as
+   *   a means for tuning job performance. The auto-inferred counts are usually much lower than the necessary counts for
+   *   Photon (especially GAME), so this caused a lot of shuffling when repartitioning from the auto-partitioned data
+   *   to the GAME data. We expose this setting here to avoid the shuffling.
+   * @return the loaded and transformed DataFrame
+   */
+  override def readMerged(
+      paths: Seq[String],
+      featureColumnMap: Map[String, Set[String]],
+      numPartitions: Int): (DataFrame, Map[String, IndexMapLoader]) = {
+
+    require(paths.nonEmpty, "No paths specified. You must specify at least one input path.")
+    require(numPartitions > 0, "Partition count must be greater than zero.")
+
+    val records = AvroUtils.readAvroFiles(sc, paths, numPartitions)
+    val indexMapLoaders = generateIndexMapLoaders(records, featureColumnMap)
+
+    (readMerged(paths, indexMapLoaders, featureColumnMap, numPartitions), indexMapLoaders)
```

Comment:
You're right, good catch. I'm going to make a new method that accepts RDD[GenericRecord] as input, so that we can reuse the RDD instead of re-reading.